lib/wickedbase.pm

```diff
@@ -117,6 +119,104 @@ sub assert_wicked_state {
     $self->ping_with_timeout(ip => $args{ping_ip}) if $args{ping_ip};
 }
 
+sub valgrind_get_services {
+    my $self = shift;
+    my $val = get_var('WICKED_VALGRIND');
+
+    return if (!defined($val) || !$val);
+    my @all = qw(wickedd-auto4
+      wickedd-dhcp6
+      wickedd
+      wickedd-dhcp4
+      wickedd-nanny
+      wicked);
+    my @tmp;
+    my @enable;
+    if (index($val, ",")) {
+        @tmp = split(/,/, $val);
+    } else {
+        @tmp = $val;
+    }
+    foreach my $v (@tmp) {
+        $v =~ s/\.service$//;
+        if (grep { /^$v$/ } @all) {
+            push @enable, $v;
+        } elsif ($v =~ /^1|all$/i) {
+            push @enable, @all;
+        } else {
+            record_info('WARNING', "Unknown value for WICKED_VALGRIND $v", result => 'softfail');
+        }
+    }
+    @enable = uniq @enable;
+
+    return @enable;
+}
+
+sub valgrind_cmd {
+    my ($self, $service) = @_;
+
+    my $valgrind_cmd = get_var('WICKED_VALGRIND_CMD', '/usr/bin/valgrind --tool=memcheck --leak-check=yes');
+    if ($service) {
+        my $logfile = "/var/log/valgrind_$service.log";
+        $valgrind_cmd = "$valgrind_cmd --log-file=$logfile";
+    }
+    return $valgrind_cmd;
+}
+
+sub valgrind_enable {
+    my $self = shift;
+    my $valgrind_cmd = get_var('WICKED_VALGRIND_CMD', '/usr/bin/valgrind --tool=memcheck --leak-check=yes');
+
+    my @services = $self->valgrind_get_services();
+    return 0 if (!@services);
+
+    record_info("valgrind enable", "services: @services\ncommand: $valgrind_cmd");
+
+    foreach my $service (@services) {
+        my $service_file = "/etc/systemd/system/$service.service";
+
+        assert_script_run("systemctl cat $service > $service_file");
+        assert_script_run(sprintf(q(sed -i -E 's@^(ExecStart=)(.*)$@\1%s \2@' '%s'),
```

Comment:
I think this line requires some comments explaining why you are sure that ExecStart will have exactly this content and maybe one example of expected content

Reply:
What we do here, we add a prefix (the valgrind command) to the executable of the service. The only thing we expect is `ExecStart=` which must be there, see: https://www.freedesktop.org/software/systemd/man/systemd.service.html#ExecStart= 

More details:
 * We use `q()` as string quotes to allow `'` and special chars like "@" in it.
 * We build a more complex string with `sprintf("%s", $var)` to make the reading of the `sed` command more easy.
 * The `sed` call replace the line starting with `ExecStart=<command>` with `ExecStart=<valgrind-cmd> <command>`
   *  we use `@` as regex delemiter, to allow `/` 
   * `\1` and `\2` in the replace are the findings of group 1 and 2, where group 1 is always `ExecStart=`

Maybe as comment:
```
# Add valgrind command prefix to `ExecStart=` in service file
```

Reply:
> Maybe as comment:

> # Add valgrind command prefix to `ExecStart=` in service file

yes please

Reply:
done